### PR TITLE
Implement authentication through a Facebook flow

### LINKED ISSUE #24: User authentication (initially with Facebook only)

### PR DESCRIPTION
**Pull request by julienw**
Closes #24

### CODE REVIEW

File: src/config.js [hunk 1]
```diff
@@ -46,12 +55,12 @@ const conf = convict({
...
     host: {
       doc: 'Hostname where PostgreSQL is running',
-      format: 'url',
+      format: '*',
       default: 'localhost'
     },
     port: {
       doc: 'Port where PostgreSQL is running',
-      format: 'port',
+      format: '*',
```
Comment (julienw) [inline src/config.js]: these changes are needed for me because I connect to psql through a unix socket (which is quite cool because it matches psql user with your user :) ).

Comment (julienw) [inline src/config.js]: but I'm fine not committing this and keeping the change locally

File: src/middlewares/auth.js [hunk 2]
```diff
@@ -16,51 +17,73 @@ function unauthorized(res) {
...
+
+  if (!validScopes.length) {
+    throw new Error(`No valid scope found in "${scopes}"`);
+  }
+
   return (req, res, next) => {
-    const authHeader = req.headers['authorization'];
-    if (!authHeader) {
+    const authHeader = req.headers.authorization;
+    const authQuery = req.query.authorizationToken;
```
Comment (julienw) [inline src/middlewares/auth.js]: To be able to use a simple browser request, we can pass the JWT token in this query parameter. I haven't looked if there is anything kinda standardized for this yet.

File: src/middlewares/auth.js [hunk 3]
```diff
@@ -16,51 +17,73 @@ function unauthorized(res) {
...
     const decoded = jwt.decode(token);
 
-    // For now we only allow authenticated requests from the admin user.
-    // When this changes we will have a different secret per sensor and per
-    // user.
-    if (!decoded || !decoded.id || decoded.id !== 'admin') {
+    if (!decoded || !decoded.id || !decoded.scope) {
+      return unauthorized(res);
+    }
+
```
Comment (julienw) [inline src/middlewares/auth.js]: note: directly calling `unauthorized` disallows the redirect to `failureUrl`; this will be fixed once we just pass an error to `next()` so that we can handle it (or not) in the next middleware.

File: src/models/db.js [hunk 4]
```diff
@@ -27,7 +27,16 @@ let deferreds = [];
...
 
-module.exports = () => {
+const { name, user, password, host, port } = config.get('db');
+
+const sequelize = new Sequelize(name, user, password, {
+  host,
+  port,
+  dialect: 'postgres',
+  logging: false
+});
```
Comment (julienw) [inline src/models/db.js]: extracting this out of the exported function makes it possible to export it so that we can use it elsewhere. (in my case: for the session store)

Comment (julienw) [inline src/models/db.js]: I don't think this makes a big perf difference, as I believe what makes the big perf hit is loading all the model and calling `sync`

File: src/models/users.js [hunk 5]
```diff
@@ -9,22 +6,65 @@ import {
...
+  BASIC: ({ username, password }) => {
+    // For now we only support admin authentication.
+    if (username !== 'admin' || password !== config.get('adminPass')) {
+      return Promise.reject(new Error(UNAUTHORIZED));
+    }
+
+    return Promise.resolve({
+      id: 'admin',
+      scope: 'admin'
+    });
```
Comment (julienw) [inline src/models/users.js]: basic authentication is for admin only

File: src/models/users.js [hunk 6]
```diff
@@ -9,22 +6,65 @@ import {
...
+  },
+
+  AUTH_PROVIDER: (data) => {
+    return Promise.resolve({
+      id: data,
+      scope: 'user'
```
Comment (julienw) [inline src/models/users.js]: auth provider based login is for users only

File: src/models/users.js [hunk 7]
```diff
@@ -9,22 +6,65 @@ import {
...
+      .then(userData => {
+        if (userData.scope !== 'user') {
+          return userData;
+        }
+
+        return User.findOrCreate({
+          attributes: [],
+          where: userData.id, // this contains all user attributes
+        })
+          .then(() => userData);
```
Comment (julienw) [inline src/models/users.js]: not sure whether I should put a _real_ user object in `userData.id`

File: src/routes/auth/basic.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+passport.use(new BasicStrategy(
+  (username, password, done) => {
+    db().then(models => {
+      const { BASIC, authenticate } = models.Users;
+      return authenticate(BASIC, { username, password });
+    }).then(
+      userInfo => done(null, userInfo),
+      err => {
+        if (err.message === UNAUTHORIZED) {
+          done(null, false);
```
Comment (julienw) [inline src/routes/auth/basic.js]: unauthorized errors means "no user", passing `false` is how passport understands this

File: src/routes/auth/facebook.js [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+  }
+);
+
+router.get(
+  '/callback',
+  checkHasValidSession,
+  (req, res, next) => {
+    passport.authenticate(
+      'facebook',
+      (err, user, _info) => {
```
Comment (julienw) [inline src/routes/auth/facebook.js]: using a callback is the only way I found to customize what to do when there is a failure

File: src/routes/auth/index.js [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+import basic    from './basic';
+import facebook from './facebook';
+
+import config   from '../../config';
+import { sequelize }       from '../../models/db';
+
+const router = express.Router();
+
+router.use('/basic', basic);
```
Comment (julienw) [inline src/routes/auth/index.js]: no session for the basic auth, that's why I use it before doing the session stuff

File: src/routes/auth/index.js [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+  secret: 'keyboard cat',
+  store: new SequelizeStore({
+    db: sequelize
+  }),
+  cookie: {
+    path: `/${config.get('version')}/auth`,
+    secure: isProduction,
+  },
+  name: 'connect.sid.auth',
+  proxy: isProduction, // TODO maybe better configure this in config
```
Comment (julienw) [inline src/routes/auth/index.js]: this is when node is in HTTP but there is a proxy in HTTPS, when we use `secure: true` above

File: src/routes/auth/utils.js [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+
+import config   from '../../config';
+
+export function finalizeAuth(req, res) {
+  const token = jwt.sign(req.user, config.get('adminSessionSecret'));
+
+  if (req.session && req.session.redirectUrl) {
+    const redirectUrl = url.parse(req.session.redirectUrl, true);
+    redirectUrl.query.token = token;
+    req.session.destroy();
```
Comment (julienw) [inline src/routes/auth/utils.js]: I think I'll need to destroy this in case of failures too... but not so easy to do with the current way of managing errors :)

**Comment by julienw** [comment]
feedback? @ferjm 

Now I need to work on tests and clean up some style :)

**Comment by julienw** [comment]
BTW I use https://github.com/julienw/sensorweb-login-website to start my login flow. Here are the informations for my client with the hardcoded JWT:
```
name: julien
key: 7c91e8d619046ad7
secret: efa0a85e98bfc7929e66a18e14db5cdae4fce6517dc48e3c138fcf3d207b94b3af883985a7134adef94790b59acdb1be220011f8aff903cab8729cbdb45005ff
authRedirectUrls: {http://localhost:10000/token.html}
authFailureRedirectUrls: {http://localhost:10000/tokenFailure.html}

File: src/models/users.js (hunk 7 above)
Comment (ferjm) [inline src/models/users.js]: It is probably safer to create a new object only with the fields that we care about (i.e., opaqueId and provider). Sequelize adds some extra fields like `createdAt` and `updatedAt` that we should remove from the response.

File: src/routes/auth/facebook.js [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+  ApiError,
+  BAD_REQUEST, ERRNO_BAD_REQUEST,
+  FORBIDDEN, ERRNO_FORBIDDEN,
+  UNAUTHORIZED, ERRNO_UNAUTHORIZED,
+} from '../../errors';
+
+const router = express.Router();
+
+const callbackURL =
+  `${config.get('publicHost')}/${config.get('version')}/auth/facebook/callback`;
```
Comment (ferjm) [inline src/routes/auth/facebook.js]: Could we move the last part also to the config, please?

File: src/routes/auth/facebook.js [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+    clientID: config.get('facebook.clientID'),
+    clientSecret: config.get('facebook.clientSecret'),
+    callbackURL,
+    passReqToCallback: true,
+    enableProof: true,
+    state: true,
+  },
+  function(req, accessToken, refreshToken, profile, cb) {
+    db()
+      .then(models => {
```
Comment (ferjm) [inline src/routes/auth/facebook.js]: nit: weird indentation

File: src/models/users.js (hunk 7 above)
Comment (ferjm) [inline src/models/users.js]: Also, uber-nit: weird indentation.

File: src/models/users.js [hunk 15]
```diff
@@ -9,22 +6,65 @@ import {
...
+};
+
+module.exports = (sequelize, DataTypes) => {
+  const User = sequelize.define('Users', {
+    opaqueId: { type: DataTypes.STRING(256) },
+    provider: { type: DataTypes.STRING(32) },
+  }, {
+    indexes: [{
+      unique: true,
+      fields: ['opaqueId', 'provider', 'ClientKey']
```
Comment (ferjm) [inline src/models/users.js]: Could you add a comment about what's `ClientKey` for? Also, does it need to be in capital letters?

File: src/routes/auth/facebook.js [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+      .then(
+        userInfo => cb(null, userInfo),
+        err => cb(err)
+      );
+  }
+));
+
+function checkClientExists(req, res, next) {
+  db()
+    .then(({ Clients }) => Clients.findById(req.user.id))
```
Comment (ferjm) [inline src/routes/auth/facebook.js]: Could we s/req.user/req.client/g?

File: config/sample.json [hunk 17]
```diff
@@ -3,13 +3,18 @@
...
   "port": 8080,
+  "publicHost": "http://localhost:8080",
   "db": {
     "host": "localhost",
     "port": 5432,
     "name": "sensorweb",
     "user": "postgres",
     "password": "default"
   },
+  "facebook": {
```
Comment (julienw) [inline config/sample.json]: sure !

File: package.json [hunk 18]
```diff
@@ -37,16 +37,24 @@
...
     "jsonwebtoken": "^7.1.9",
     "lodash": "^4.15.0",
     "morgan-body": "^0.9.1",
     "on-headers": "^1.0.1",
     "owasp-password-strength-test": "^1.3.0",
+    "passport": "^0.3.2",
+    "passport-facebook": "^2.1.1",
+    "passport-google-oauth": "^1.0.0",
+    "passport-http": "^0.3.0",
+    "passport-twitter": "^1.0.4",
```
Comment (julienw) [inline package.json]: ooops right :)

File: src/config.js [hunk 19]
```diff
@@ -20,14 +20,23 @@ const password = value => {
...
+  }
+});
+
 const conf = convict({
   adminPass: {
     doc: 'The password for the admin user. Follow OWASP guidelines for passwords',
     format: password,
     default: 'invalid'
   },
   adminSessionSecret: {
```
Comment (julienw) [inline src/config.js]: yeah, we discussed to do it in a separate patch because it's orthogonal and I didn't want to cram too much things into this patch. But I should probably revert the next change then.

File: src/config.js (hunk 1 above)
Comment (julienw) [inline src/config.js]: I think it's probably better to do something in a separate patch; possibly with a custom verifier. I'll revert this from this patch for now.

File: src/middlewares/auth.js [hunk 20]
```diff
@@ -16,51 +17,73 @@ function unauthorized(res) {
...
-      return unauthorized(res);
+    let token;
+    if (authHeader) {
+      const token = authHeader.split('Bearer ')[1];
+      if (!token) {
+        return unauthorized(res);
+      }
+    } else {
+      token = authQuery;
     }
```
Comment (julienw) [inline src/middlewares/auth.js]: yeah why not; I can use a ternary which can be a tiny bit more readable.

File: src/models/users.js (hunk 15 above)
Comment (julienw) [inline src/models/users.js]: It's created automatically by sequelize with the association we create in `associate`.

I agree this is not really reader-friendly because the association is done quite a few lines below. A comment will definitely help here.

I can also configure explicitely the foreign key so that we control the name, and then we can have a lowercase as first letter. Do you think it's better even with the comment ?

File: src/models/users.js (hunk 7 above)
Comment (julienw) [inline src/models/users.js]: Note that the result to `authenticate` is not returned to the user, rather it's stored in `req.user` for the next (express) middlewares.

Currently I'm not returning the sequelize's user at all. `userData` comes from the authentication methods. Depending on the methods we don't always have the same structure. My take at this is that the "scope" will give us the structure.

* `userData.scope === 'user'` => `userData.id` is an object with `opaqueId`, `provider` and `ClientKey`. This looks like the sequelize object but it's not. It comes from the authentication methods to we rely on them giving us the right structure (but a wrong structure would make the find fail, so I think we're good).

  My main issue with the current approach is that in this function we never know explicitely what's inside `userData.id`, that could make it more difficult to read the function.

* `userData.scope === 'admin'` => `userData.id` is `'admin'`
* `userData.scope === 'client'` => `userData.id` is just the client's key

So once we know the `scope` we also know how to consult the `id`.

Comment (julienw) [inline src/models/users.js]: Note: the indentation comes from my work with gmarty :) each `.then` go a new line even if it was not completely needed. I liked the fact it was a systematic rule so I kind of adopted it too, but I can also adapt myself to another code style :D

File: src/routes/auth/facebook.js (hunk 13 above)
Comment (julienw) [inline src/routes/auth/facebook.js]: you mean the `callback` part ? In that case we can make `callback` the default ?

Just wondering: why do you think it's needed ?

File: src/routes/auth/basic.js (hunk 8 above)
Comment (julienw) [inline src/routes/auth/basic.js]: sure

File: src/routes/auth/facebook.js (hunk 16 above)
Comment (julienw) [inline src/routes/auth/facebook.js]: actually, it seems to be a standard practice in express apps that the result of the authentications is in `req.user`. Also `req.user` can contains the information for clients, users or admin (and later: sensors).

The alternative is that instead of having a `scope` we could either set `req.user` `req.client` `req.admin` and `req.sensor` with the information we currently have in `id`. This looks quite more elegant to me. What do you think ?

File: src/routes/auth/index.js [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+const router = express.Router();
+
+router.use('/basic', basic);
+
+// initalize sequelize with session store
+const SequelizeStore = SequelizeStoreFactory(session.Store);
+const isProduction = process.env.NODE_ENV === 'production';
+router.use(cookieParser());
+router.use(session({
+  secret: 'keyboard cat',
```
Comment (julienw) [inline src/routes/auth/index.js]: yeah, I thought of this too and then forgot :)

File: src/routes/auth/utils.js [hunk 22]
```diff
@@ -0,0 +1,18 @@
+import jwt      from 'jsonwebtoken';
```
Comment (julienw) [inline src/routes/auth/utils.js]: good idea !

File: src/server.js [hunk 23]
```diff
@@ -11,7 +11,7 @@ import config           from './config';
 
 import clients          from './routes/clients';
 import dockerflow       from './routes/dockerflow';
-import users            from './routes/users';
+import authRoute        from './routes/auth';
```
Comment (julienw) [inline src/server.js]: sure

File: src/routes/auth/index.js (hunk 10 above)
Comment (julienw) [inline src/routes/auth/index.js]: sure

File: src/models/users.js (hunk 15 above)
Comment (ferjm) [inline src/models/users.js]: If it is not too much effort, configuring the fk explicitly sounds good. It's not a big deal though.

File: src/routes/auth/facebook.js (hunk 13 above)
Comment (ferjm) [inline src/routes/auth/facebook.js]: Yes, basically all the hard-coded part. Having it in the config, in a common place for all IdP configuration, makes things a little bit easier, specially when trying to develop locally.

File: src/routes/auth/facebook.js (hunk 16 above)
Comment (ferjm) [inline src/routes/auth/facebook.js]: Yes, this looks more elegant. Thanks.

Comment (ferjm) [inline src/routes/auth/facebook.js]: Also, just to be extra careful, could you exclude the client secret from the query, please?

File: src/models/users.js (hunk 7 above)
Comment (ferjm) [inline src/models/users.js]: Ok. Thanks for the explanation.

Regarding the indentation, it's not a big deal, but rather that introducing a new one, try to follow the existing style, please.

```javascript
return User.findOrCreate({
...
}).then(...
```
or even
```javascript
return User.findOrCreate({
...
})
.then(...
```

File: src/config.js [hunk 24]
```diff
@@ -93,6 +105,37 @@ const conf = convict({
       }
     }
   },
+  behindProxy: {
+    doc: `Set this to true if the server runs behind a reverse proxy. This is
+          especially important if the proxy implements HTTPS with
+          userAuth.cookieSecure. Also with this Express will trust the
+          X-Forwarded-For header. Set to 1 or auto if you're behind a proxy.`,
+    default: false,
+    format: '*',
```
Comment (julienw) [inline src/config.js]: format is "*" because otherwise convict infers it's a boolean from the default and will refuse 1 or "auto".

File: src/middlewares/auth.js [hunk 25]
```diff
@@ -31,36 +43,43 @@ export default (scopes) => {
...
+        if (error) {
+          console.log('Error while verifying the token', error);
+          return unauthorized(res);
+        }
 
-      // For now we only allow 'admin' scope.
-      const validScopes = ['admin'].filter(scopeIndex => {
-        return scopes.indexOf(scopeIndex) != -1;
+        req[decoded.scope] = decoded.id;
+        req.authScope = decoded.scope;
```
Comment (julienw) [inline src/middlewares/auth.js]: these 2 lines are repeated a few times, it's likely a good idea to factorize this, but maybe in a separate patch ?

File: src/server.js [hunk 26]
```diff
@@ -29,7 +31,7 @@ app.use(expressValidator({
 }));
 app.use(bodyParser.json());
 
-if (config.get('env') !== 'test') {
+if (config.get('env') !== 'test' || process.env.FORCE_OUTPUT) {
```
Comment (julienw) [inline src/server.js]: to be able to see the output in tests
The alternative is to have an env like "QUIET=1" to remove the logger, and set it in travis.

That said, I usually prefer to have more log than less log :)

File: test/common.js [hunk 27]
```diff
@@ -0,0 +1,36 @@
+import co  from 'co';
+import jwt from 'jsonwebtoken';
+
+import config from '../src/config';
+
+const endpointPrefix = `/${config.get('version')}`;
+
+export function loginAsAdmin(server) {
+  return co(function*() {
```
Comment (julienw) [inline test/common.js]: if you're not familiar with `co`, it returns a Promise and takes a generator as argument; this makes it easier to deal with Promise-returning functions. `yield` will wait for the Promise to be resolved, rejection translates to exceptions, etc. This is so good :)

Comment (julienw) [inline test/common.js]: Here we could just use promise chains because it stays quite simple, still I think it's more readable like this.

File: test/mocha.opts [hunk 28]
```diff
@@ -0,0 +1,2 @@
+--require co-mocha
```
Comment (julienw) [inline test/mocha.opts]: This is a sugar helper to wrap the mocha tests in co if it's a generator.

File: test/test_auth_api.js [hunk 29]
```diff
@@ -0,0 +1,228 @@
...
+      server.post(endpointPrefix + '/auth/basic')
+            .set('Authorization', 'Basic ' + pass)
+            .expect(200)
+            .end((err, res) => {
+              res.status.should.be.equal(201);
+              should.exist(res.body.token);
+              done();
+            });
+    });
+  });
```
Comment (julienw) [inline test/test_auth_api.js]: I only grouped them inside the same `describe` + the endpoint, no other change. There are few things we could do though: template strings, returning a promise instead of using done, etc.

File: test/test_auth_api.js [hunk 30]
```diff
@@ -0,0 +1,228 @@
...
+} from '../src/errors';
+
+import {
+  loginAsAdmin,
+  loginAsClient,
+  createClient,
+} from './common';
+
+const endpointPrefix = '/' + config.get('version');
+const server = supertest(app);
```
Comment (julienw) [inline test/test_auth_api.js]: it's not an agent anymore; an agent keeps state between calls and I think we should do it only when needed.

File: test/test_clients_api.js [hunk 31]
```diff
@@ -24,7 +24,7 @@ let token;
 describe('Clients API', () => {
   before((done) => {
     const pass = btoa('admin:' + config.get('adminPass'));
-    server.post(endpointPrefix + '/users/auth')
+    server.post(endpointPrefix + '/auth/basic')
```
Comment (julienw) [inline test/test_clients_api.js]: this "before" could use the common file now, but maybe later in a separate patch

File: test/test_users_api.js [hunk 32]
```diff
@@ -1,76 +0,0 @@
-import btoa       from 'btoa';
```
Comment (julienw) [inline test/test_users_api.js]: renamed to test_auth_api

File: package.json [hunk 33]
```diff
@@ -26,27 +26,36 @@
...
     "coveralls": "^2.11.14",
     "istanbul": "^0.4.5",
     "mocha": "^3.1.0",
+    "nock": "^9.0.2",
     "nodemon": "^1.10.2",
     "nyc": "^8.3.1",
     "should": "^11.1.0",
-    "supertest": "^2.0.0"
+    "supertest": "^2.0.0",
+    "supertest-as-promised": "^4.0.2"
```
Comment (julienw) [inline package.json]: this is nice as we can work with promises instead of using end + done callback.

**Comment by julienw** [comment]
I haven't checked if travis/circleci works yet. I think circleci doesn't because I haven't adapted the config file yet.

also there is a breaking change: the endpoint to login changed, so we'll need to update the admin panel.

**Comment by julienw** [comment]
CircleCi is green now, so @ferjm r? 

If you review only the changes from last time, please note I reverted some of the lint changes I did in the first patch. So if you see semicolons disappearing, dont be surprised :)

File: test/test_clients_api.js (hunk 31 above)
Comment (ferjm) [inline test/test_clients_api.js]: Could you file an issue and add a XXX comment here referencing it, please?

File: test/test_auth_api.js (hunk 29 above)
Comment (ferjm) [inline test/test_auth_api.js]: File issues for these suggestions and reference them in the code, please.

File: test/test_auth_api.js [hunk 34]
```diff
@@ -0,0 +1,228 @@
...
+      const agent = supertest.agent(app);
+      let res = yield agent.get(endpoint)
+                           .query({ authorizationToken: clientToken })
+                           .query({ redirectUrl: redirectUrls[1] })
+                           .query({ failureUrl: failureRedirectUrls[1] })
+                           .expect(302)
+                           .expect('location', /facebook\.com/)
+                           .expect('set-cookie', /^connect\.sid\.auth=/);
+
+      // This is Faacebook's anti-CSRF protection
```
Comment (ferjm) [inline test/test_auth_api.js]: typo: Facebook

File: src/middlewares/auth.js (hunk 25 above)
Comment (julienw) [inline src/middlewares/auth.js]: there are 3 locations where we set `req[XXX.scope]` and `req.authScope`: here, and in all `auth/<method>.js`  (basic, facebook, and later google and fxaccounts likely).

That said, maybe we don't want this in `auth/<method>.js` where it's only consumed by `finalizeAuth`; we could simply use req.user or even something else like req.auth.

I'll think of this and maybe we can rework it when implementing google and fx accounts.

File: src/server.js (hunk 26 above)
Comment (ferjm) [inline src/server.js]: I'm fine with both options as long as we have a way to keep the logging quite.

File: src/middlewares/auth.js (hunk 25 above)
Comment (ferjm) [inline src/middlewares/auth.js]: File another issue and reference it in the code, please :)

Comment (julienw) [inline src/middlewares/auth.js]: after all I just did what I suggested in a separate commit. Easier than creating an issue :) Tell me what you think.
